In [30]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

### 중국 주가 데이터 읽기 및 확인

In [32]:
# 중국의 주가 데이터
cn_a50 = pd.read_csv("./중국사회경제/중국A50.csv")
cn_szse = pd.read_csv("./중국사회경제/중국SZSE.csv")
cn_sha = pd.read_csv("./중국사회경제/중국상해종합.csv")

In [33]:
cn_a50.head()

날짜        종가        시가        고가        저가   변동 %
0  2024-05-17  12793.29  12629.94  12793.29  12626.83  129.0
1  2024-05-16  12629.94  12532.82  12675.55  12526.58   77.0
2  2024-05-15  12532.82  12603.21  12635.08  12524.33  -56.0
3  2024-05-14  12603.21  12644.30  12672.37  12569.84  -32.0
4  2024-05-13  12644.30  12697.15  12697.15  12561.24  -42.0

In [34]:
cn_a50.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2299 entries, 0 to 2298
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   날짜      2299 non-null   object 
 1   종가      2299 non-null   float64
 2   시가      2299 non-null   float64
 3   고가      2299 non-null   float64
 4   저가      2299 non-null   float64
 5   변동 %    2299 non-null   float64
dtypes: float64(5), object(1)
memory usage: 107.9+ KB


In [35]:
cn_szse.head()

날짜       종가       시가       고가       저가           거래량   변동 %
0  2024-05-17  9709.42  9590.39  9709.42  9563.47  1.527000e+09  110.0
1  2024-05-16  9604.13  9614.73  9668.85  9582.22  1.421000e+09   21.0
2  2024-05-15  9583.54  9648.43  9659.04  9577.21  1.267000e+09  -88.0
3  2024-05-14  9668.73  9689.55  9736.29  9633.80  1.341000e+09   -5.0
4  2024-05-13  9673.32  9659.64  9720.48  9605.18  1.379000e+09  -60.0

In [36]:
cn_szse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2277 entries, 0 to 2276
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   날짜      2277 non-null   object 
 1   종가      2277 non-null   float64
 2   시가      2277 non-null   float64
 3   고가      2277 non-null   float64
 4   저가      2277 non-null   float64
 5   거래량     2277 non-null   float64
 6   변동 %    2277 non-null   float64
dtypes: float64(6), object(1)
memory usage: 124.7+ KB


In [37]:
cn_sha.head()

날짜       종가       시가       고가       저가           거래량   변동 %
0  2024-05-17  3154.03  3121.71  3154.03  3116.90  3.839200e+05  101.0
1  2024-05-16  3122.40  3127.18  3137.88  3116.16  3.674000e+09    8.0
2  2024-05-15  3119.92  3141.25  3142.52  3117.93  3.171200e+05  -84.0
3  2024-05-14  3146.29  3148.85  3155.64  3137.66  3.405100e+05   -6.0
4  2024-05-13  3148.17  3139.48  3157.37  3126.17  3.931400e+05  -20.0

In [38]:
cn_sha.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2277 entries, 0 to 2276
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   날짜      2277 non-null   object 
 1   종가      2277 non-null   float64
 2   시가      2277 non-null   float64
 3   고가      2277 non-null   float64
 4   저가      2277 non-null   float64
 5   거래량     2277 non-null   float64
 6   변동 %    2277 non-null   float64
dtypes: float64(6), object(1)
memory usage: 124.7+ KB


- 상해와 szse의 데이터 수량은 2277인데 a50의 경우 2299로 22개 더 많은 것이 확인됨

### 결측일 확인 및 불필요한 날짜 제거

In [39]:
# cn_a50에는 있으나 cn_sha는 없는 날짜 확인
cn_a50[~cn_a50["날짜"].isin(cn_sha["날짜"])]

날짜        종가        시가        고가        저가    변동 %
943   2020-06-25  13938.63  13938.63  13938.63  13938.63     0.0
980   2020-05-01  13447.92  13447.92  13447.92  13447.92     0.0
999   2020-04-06  12698.12  12698.12  12698.12  12698.12     0.0
1045  2020-01-24  13791.36  13791.36  13791.36  13791.36     0.0
1062  2020-01-01  14383.81  14383.81  14383.81  14383.81     0.0
1135  2019-09-13  13979.67  13979.67  13979.67  13979.67     0.0
1205  2019-06-07  12622.15  12622.15  12622.15  12622.15     0.0
1369  2018-10-01  11873.28  11873.28  11873.28  11873.28     0.0
1443  2018-06-18  12482.79  12482.79  12482.79  12482.79     0.0
1477  2018-04-30  12192.25  12192.25  12192.25  12192.25     0.0
1493  2018-04-05  12991.75  12991.75  12991.75  12991.75   377.0
1524  2018-02-15  13353.50  13353.50  13353.50  13353.50     0.0
1557  2018-01-01  13195.72  13195.72  13195.72  13195.72     0.0
1706  2017-05-30  10992.12  10992.12  10992.12  10992.12     0.0
1707  2017-05-29  10992.12  10992.12  10992.12  10992.12     0.0
1727  2017-05-01  10390.58  10390.58  10390.58  10390.58     0.0
1746  2017-04-03  10407.13  10407.13  10407.13  10407.13     0.0
1806  2017-01-01  13195.72  13195.72  13195.72  13195.72  3235.0
1867  2016-10-07   9598.73   9598.73   9598.73   9598.73     0.0
1868  2016-10-03   9598.73   9598.73   9598.73   9598.73     0.0
1879  2016-09-15   9557.45   9557.45   9557.45   9557.45     0.0
2298  2015-01-04  11517.68  11517.68  11517.68  11517.68     0.0

In [23]:
# cn_a50에는 있으나 cn_szse에는 없는 날짜 확인
cn_a50[~cn_a50["날짜"].isin(cn_szse["날짜"])]

날짜        종가        시가        고가        저가    변동 %
943   2020-06-25  13938.63  13938.63  13938.63  13938.63     0.0
980   2020-05-01  13447.92  13447.92  13447.92  13447.92     0.0
999   2020-04-06  12698.12  12698.12  12698.12  12698.12     0.0
1045  2020-01-24  13791.36  13791.36  13791.36  13791.36     0.0
1062  2020-01-01  14383.81  14383.81  14383.81  14383.81     0.0
1135  2019-09-13  13979.67  13979.67  13979.67  13979.67     0.0
1205  2019-06-07  12622.15  12622.15  12622.15  12622.15     0.0
1369  2018-10-01  11873.28  11873.28  11873.28  11873.28     0.0
1443  2018-06-18  12482.79  12482.79  12482.79  12482.79     0.0
1477  2018-04-30  12192.25  12192.25  12192.25  12192.25     0.0
1493  2018-04-05  12991.75  12991.75  12991.75  12991.75   377.0
1524  2018-02-15  13353.50  13353.50  13353.50  13353.50     0.0
1557  2018-01-01  13195.72  13195.72  13195.72  13195.72     0.0
1706  2017-05-30  10992.12  10992.12  10992.12  10992.12     0.0
1707  2017-05-29  10992.12  10992.12  10992.12  10992.12     0.0
1727  2017-05-01  10390.58  10390.58  10390.58  10390.58     0.0
1746  2017-04-03  10407.13  10407.13  10407.13  10407.13     0.0
1806  2017-01-01  13195.72  13195.72  13195.72  13195.72  3235.0
1867  2016-10-07   9598.73   9598.73   9598.73   9598.73     0.0
1868  2016-10-03   9598.73   9598.73   9598.73   9598.73     0.0
1879  2016-09-15   9557.45   9557.45   9557.45   9557.45     0.0
2298  2015-01-04  11517.68  11517.68  11517.68  11517.68     0.0

In [40]:
# 결측된 두 날짜의 리스트가 서로 동일하다는 것을 확인
cn_a50[~cn_a50["날짜"].isin(cn_sha["날짜"])].index == cn_a50[~cn_a50["날짜"].isin(cn_szse["날짜"])].index

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True])

In [41]:
# 결측 날짜의 길이 확인
len(cn_a50[~cn_a50["날짜"].isin(cn_sha["날짜"])])

22

- 결측일 대부분의 변동률이 0%인 것으로 확인됨
- 세부조사 결과 변동률이 0%인 날은 주말로 불필요한 데이터
- 단 변동률이 있는 2일은 cn_sha와 cn_szse측 결측으로 생각됨
    - interpolate를 사용하여 처리 예정

In [42]:
# 결측일 중 변동률이 0인 날짜들의 인덱스 특정
drop_idx = cn_a50[(~cn_a50["날짜"].isin(cn_sha["날짜"])) & (cn_a50["변동 %"] == 0)].index

In [43]:
# drop_idx를 cn_a50에서 제거
cn_a50 = cn_a50.drop(drop_idx, axis = 0).reset_index(drop = True)

### 병합 작업

In [44]:
# 각 데이터프레임에서 날짜와 종가를 제외한 나머지 피처 제거
cn_a50 = cn_a50.loc[:, ["날짜", "종가"]]
cn_szse = cn_szse.loc[:, ["날짜", "종가"]]
cn_sha = cn_sha.loc[:, ["날짜", "종가"]]

In [49]:
# 각 종가 피처의 명칭을 데이터프레임의 명칭으로 변경
cn_a50.columns = ["날짜", "cn_a50"]
cn_szse.columns = ["날짜", "cn_szse"]
cn_sha.columns = ["날짜", "cn_sha"]

In [83]:
# cn_stock이라는 데이터프레임으로 병합
cn_stock = pd.merge(cn_a50, cn_szse, on = "날짜", how = "outer")
cn_stock = pd.merge(cn_stock, cn_sha, on = "날짜", how = "outer" )

### 결측치 제거

In [84]:
# 결측치 확인
cn_stock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2279 entries, 0 to 2278
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   날짜       2279 non-null   object 
 1   cn_a50   2279 non-null   float64
 2   cn_szse  2277 non-null   float64
 3   cn_sha   2277 non-null   float64
dtypes: float64(3), object(1)
memory usage: 71.3+ KB


- cn_szse와 cs_sha에 2개씩의 결측이 있는 것이 확인됨

In [85]:
# 2개의 결측 데이터 확인
cn_stock[cn_stock["cn_szse"].isna()]

날짜    cn_a50  cn_szse  cn_sha
1483  2018-04-05  12991.75      NaN     NaN
1790  2017-01-01  13195.72      NaN     NaN

In [86]:
# interpolate("linear")를 활용하여 2개의 결측치 제거
cn_stock = cn_stock.interpolate("linear")

C:\Users\ITSC\AppData\Local\Temp\ipykernel_8064\3538828065.py:2: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  cn_stock = cn_stock.interpolate("linear")


### 주말, 공휴일 추가

- 주말과 공휴일엔 거래가 없는 주식 특성 상 해당 데이터는 자연스럽게 결측되어 있음
    - 15년 1월 1일부터 24년 5월 19일까지는 3427일로 약 1/3의 데이터가 결측된 상태임
    
- 주말과 공휴일의 경우 ffill()을 활용해 처리하기로 한다

In [87]:
# 날짜 피처를 datetime으로 변환
cn_stock["날짜"] = pd.to_datetime(cn_stock["날짜"])

In [88]:
# 날짜 피처를 인덱스로 설정
cn_stock = cn_stock.set_index("날짜")

In [89]:
# 모든 날짜를 보유한 리스트 생성
date_list = pd.date_range(start = "2015-01-01", end = "2024-05-19")

In [98]:
# cn_stock에 없는 일시에 대해서만 np.nan 값을 지니는 데이터 생성
for date in date_list:
    if date not in cn_stock.index:
        cn_stock.loc[date, :] = np.nan

In [102]:
# 날짜(index)를 기준으로 정렬
cn_stock = cn_stock.sort_index()

In [104]:
# ffill 기법으로 결측치 처리
cn_stock = cn_stock.ffill()

In [110]:
cn_stock

cn_a50   cn_szse   cn_sha
날짜                                     
2015-01-01       NaN       NaN      NaN
2015-01-02       NaN       NaN      NaN
2015-01-03       NaN       NaN      NaN
2015-01-04       NaN       NaN      NaN
2015-01-05  11860.36  11520.59  3350.52
...              ...       ...      ...
2024-05-15  12532.82   9583.54  3119.92
2024-05-16  12629.94   9604.13  3122.40
2024-05-17  12793.29   9709.42  3154.03
2024-05-18  12793.29   9709.42  3154.03
2024-05-19  12793.29   9709.42  3154.03

[3427 rows x 3 columns]

- ffill로는 1월 1일부터 1월 4일까지의 데이터를 처리할 수 없음
- bfill로 4일간의 데이터 처리

In [111]:
# bfill 적용
cn_stock = cn_stock.bfill()

In [112]:
cn_stock

cn_a50   cn_szse   cn_sha
날짜                                     
2015-01-01  11860.36  11520.59  3350.52
2015-01-02  11860.36  11520.59  3350.52
2015-01-03  11860.36  11520.59  3350.52
2015-01-04  11860.36  11520.59  3350.52
2015-01-05  11860.36  11520.59  3350.52
...              ...       ...      ...
2024-05-15  12532.82   9583.54  3119.92
2024-05-16  12629.94   9604.13  3122.40
2024-05-17  12793.29   9709.42  3154.03
2024-05-18  12793.29   9709.42  3154.03
2024-05-19  12793.29   9709.42  3154.03

[3427 rows x 3 columns]

In [113]:
# 결측치가 완전히 제거된 것을 확인
cn_stock.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 3427 entries, 2015-01-01 to 2024-05-19
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   cn_a50   3427 non-null   float64
 1   cn_szse  3427 non-null   float64
 2   cn_sha   3427 non-null   float64
dtypes: float64(3)
memory usage: 107.1 KB


In [119]:
# csv 파일로 저장 후 확인
cn_stock.to_csv("./중국사회경제/중국주가종합.csv")
cn_stock2 = pd.read_csv("./중국사회경제/중국주가종합.csv", index_col = 0)
cn_stock2.head()

cn_a50   cn_szse   cn_sha
날짜                                     
2015-01-01  11860.36  11520.59  3350.52
2015-01-02  11860.36  11520.59  3350.52
2015-01-03  11860.36  11520.59  3350.52
2015-01-04  11860.36  11520.59  3350.52
2015-01-05  11860.36  11520.59  3350.52